### ISSUE TITLE
Generating Abstract Patterns with TensorFlow

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#1355]
Generating Abstract Patterns with TensorFlow<br>
<br>
https://ift.tt/1SXzHT2<br>
<br>
<br>
<br>

<div>
<section><p>
</p>
<center>
<a href="http://blog.otoro.net/assets/20160325/tanh_01.jpeg">
<img src="http://blog.otoro.net/assets/20160325/tanh_01.jpeg">
</a>
</center>

<p>
</p>
<center>
<i>A Diamond is Forever (2016)</i>
<br>
<code>
<br>
<a target="_blank" href="http://otoro.net/ml/netart/">Web Demo</a>
</code>
</center>

<p>This is the first in a series of posts exploring <a href="https://en.wikipedia.org/wiki/Compositional_pattern-producing_network">Compositional Pattern-Producing Networks</a> in TensorFlow.  I made the code available on a <a href="https://github.com/hardmaru/cppn-tensorflow">github</a> repo for reference.  It may seem overkill to implement CPPNs with TensorFlow when <code>numpy</code> would do the job, but we will build onto this work later on.</p>

<h2 id="introduction">Introduction</h2>

<p>In recent neural network based image generation techniques, typically the generator network will attempt to draw the entire image at once.  For example, if the desired resolution of an output image is <code>256x256</code>, then the final layer of the neural network will have <code>65536</code> values for a black and white image.  It may not be feasible to increase the output resolution to modern image resolutions (<code>2880x1800</code>) due to limitations of memory and scalability these algorithms.</p>

<p>In this post I will describe a really simple way to get neural nets to generate a high resolution image.  Rather than generating every pixel at once, the neural network’s job is to generate an individual pixel’s intensity or colour, given the location of the pixel.  Then, the entire image can be generated by just querying the network once for each pixel in the desired output image.  This method can be used to generate very high resolution images, limited by memory, provides an elegant way to allow one to zoom in and out of images, and share some fractal like properties.</p>

<p>There has been some demos on the web experimenting this technique, including karpathy’s convnetjs <a href="http://cs.stanford.edu/people/karpathy/convnetjs/demo/image_regression.html">drawing demo</a>, and javascript implementations of generative <a href="http://blog.otoro.net/2015/06/19/neural-network-generative-art/">neural network art</a> and <a href="http://blog.otoro.net/2015/07/31/neurogram/">neurogram</a>.</p>

<p>In this post I will describe how to implement this simple technique to generate random abstract art in TensorFlow.  The reason I wanted to implement this in TensorFlow is so that it can be used as a base for more interesting work in the future that take advantage of TensorFlow’s machine learning capabilities, so that this type of generative network can be used to create non-random images as well, and I will explain this in future posts.</p>

<h2 id="compositional-pattern-producing-network-description">Compositional Pattern Producing Network Description</h2>

<p>
</p>
<center>
<img src="http://blog.otoro.net/assets/20160325/cppn_schematic.svg">
<br>
<i>Schematic for our customised CPPN</i>
</center>

<p>I have explained the basics of CPPN in this earlier <a href="http://blog.otoro.net/2015/07/31/neurogram/">post</a> when CPPN is combined with the NEAT algorithm to evolve the architecture of the generative neural network.</p>

<p>In essence, CPPN is just a function, , that defines the intensity of the image for every point in space.  This makes them attractive to generate very high resolution images, as you can just call this function to get the colour or intensity of every pixel, given that pixel’s location.</p>

<p> is just a function that can be built from many mathematical operations.  It can also be represented by a neural network, with a set of weights connecting the activation gates that will be held constant when drawing an image.  So the entire image can be defined as </p>

<p>Our TensorFlow implementation will deviate a bit from the previous work done with CPPN-NEAT.  Like the previous work, our function  will return either a single real number between zero and one, to define the intensity of the image at that point (result will be a greyscale image), or a three dimensional vector, each value between , to represent colour intensities (Red, Green, Blue).  In addition, similar to CPPN-NEAT, to make the image more interesting, we will also pass in a radius term for each point as the input (defined as ), so the CPPN function will be .  The weights  of the neural network will be initialised to a random value from the unit gaussian distribution.</p>

<p>Unlike CPPN-NEAT, the neural network we will be using in this implementation will just be a multi-layered feed forward network, defined by the user.  For example, the user can modify the TensorFlow code so that  is a feed forward neural network defined by hyperbolic tangent functions, relu’s, softplus, sinusoids, etc.  Each layer will also have an extra bias input, omitted in the diagram for clarity.</p>

<p>We will also add an extra input into the CPPN function called a latent vector, , which is a vector of  real numbers (where  is typically much smaller than the total number of weighted connections in the network), so our generative network is defined as .  By modifying the values of , our network will generate different images, and the entire space of images that can be possibly generated by our network by changing  around is call the latent space.  In some sense,  can be interpreted as a compressed description of the final image, summarised with  real numbers.  If we modify  by only a small amount, as the network is a continuous function, the output image would also change around only slightly, so we can also visualise how one image in a latent space can slowly morph into another image in the same latent space, by generating images with a latent vector  that gradually moves from  to .</p>

<p>Note that empty layer in the diagram is just a preprocessing layer where a scaling factor will be applied to , , , and  before entering into the first hidden layer.</p>

<p>One argument as to why we need a latent vector as an input, when we can just make gradual adjustments to the weights to get different output images.  Practically speaking, there can be in the order of hundreds or thousands of weights in a complicated generative network, and in many generative applications, we want to keep the number of latent vectors smallish.  When we eventually train on large datasets, the latent vector can control not only the object being drawn, but also the specific style of the image.  Using some tools of probability theory, we can even force  to have nice properties, such as being independent and unit gaussian.  For example, a subset of  can represent the gender of the person, and another subset can represent the mood of the person’s face being drawn by the generative algorithm.  In subsequent posts, we will attempt to train this type of CPPN to write numbers, and maybe draw faces, animals, cars, toilets, etc.</p>

<h2 id="exploring-the-latent-space-for-a-neural-network">Exploring the Latent Space for a Neural Network</h2>

<p>The CPPN model and code required to use it to generate .png and .gif images is available on <a href="https://github.com/hardmaru/cppn-tensorflow">github</a>.  I will use that code as a reference point to generate images with an IPython session.</p>

<p>Inside the repo, <code>model.py</code> contains the <code>CPPN</code> class that generates images using TensorFlow library routines.  The <code>generator</code> method will be of most interest to experiment with modifying around the CPPN’s neural net architecture, and try some alternative architectures have been commented out there for you to play with.  We will use the <code>Sampler</code> class inside <code>sampler.py</code> which will allow us to interactively query images inside an IPython session.</p>

<p>Inside an IPython session, we first go into the repo’s directory and run</p>

<figure>

<figure><pre><code data-lang="python"><table><tbody><tr>
<td><pre>1
2</pre></td>
<td><pre><span>%</span><span>run</span> <span>-</span><span>i</span> <span>sampler</span><span>.</span><span>py</span>
<span>sampler</span> <span>=</span> <span>Sampler</span><span>(</span><span>z_dim</span> <span>=</span> <span>8</span><span>,</span> <span>scale</span> <span>=</span> <span>10.0</span><span>,</span> <span>net_size</span> <span>=</span> <span>32</span><span>)</span><span>
</span></pre></td>
</tr></tbody></table></code></pre></figure>

</figure>

<p>Sampler would create a CPPN model that uses latent vectors consisting of 8 real numbers, a scaling of 10x, and contains 32 activations at each neural net layer.  Note that these are the default values so we can also call <code>sampler = Sampler()</code> next time.</p>

<p>To generate an image, we need to generate a random latent vector  containing 8 numbers</p>

<figure>

<figure><pre><code data-lang="python"><table><tbody><tr>
<td><pre>1</pre></td>
<td><pre><span>z1</span> <span>=</span> <span>sampler</span><span>.</span><span>generate_z</span><span>()</span><span>
</span></pre></td>
</tr></tbody></table></code></pre></figure>

</figure>

<p>Afterwards, we can pass  into the generator function to see the output image.  <code>img_data</code> is just a numpy array containing the image data.</p>

<figure>

<figure><pre><code data-lang="python"><table><tbody><tr>
<td><pre>1
2</pre></td>
<td><pre><span>img_data</span> <span>=</span> <span>sampler</span><span>.</span><span>generate</span><span>(</span><span>z1</span><span>)</span>
<span>sampler</span><span>.</span><span>show_image</span><span>(</span><span>img_data</span><span>)</span><span>
</span></pre></td>
</tr></tbody></table></code></pre></figure>

</figure>

<h4 id="output">Output</h4>

<center>

</center>

<p>If we want to reset and randomise the neural network’s weights, we can call the <code>reinit()</code> method</p>

<figure>

<figure><pre><code data-lang="python"><table><tbody><tr>
<td><pre>1
2</pre></td>
<td><pre><span>sampler</span><span>.</span><span>reinit</span><span>()</span>
<span>sampler</span><span>.</span><span>show_image</span><span>(</span><span>sampler</span><span>.</span><span>generate</span><span>(</span><span>z1</span><span>))</span><span>
</span></pre></td>
</tr></tbody></table></code></pre></figure>

</figure>

<h4 id="output-1">Output</h4>

<center>

</center>

<p>Note that as we have changed the weights of the network, the latent space has changed and we will see a different output despite passing in the same  vector.</p>

<p>We can save files using <code>sampler.save_png(img_data, 'output.png')</code>.</p>

<h3 id="traversing-the-latent-space">Traversing the Latent Space</h3>

<p>Let’s generate another random image in the same latent space</p>

<figure>

<figure><pre><code data-lang="python"><table><tbody><tr>
<td><pre>1
2</pre></td>
<td><pre><span>z2</span> <span>=</span> <span>sampler</span><span>.</span><span>generate_z</span><span>()</span>
<span>sampler</span><span>.</span><span>show_image</span><span>(</span><span>sampler</span><span>.</span><span>generate</span><span>(</span><span>z2</span><span>))</span><span>
</span></pre></td>
</tr></tbody></table></code></pre></figure>

</figure>

<h4 id="output-2">Output</h4>

<center>

</center>

<p>Both images belong to the same latent space, and can be represented compactly in a vector format.  If we gradually move from  to  in little steps, and generate an image at each step, we can see how the image defined by  gets morphed into the image by .  I have created a method that creates a <code>.gif</code> file to do this.  This method has some adjustable settings to tune the timing and size of frames.  The default settings will generate a <code>.gif</code> file of around 5 megabytes.</p>

<figure>

<figure><pre><code data-lang="python"><table><tbody><tr>
<td><pre>1</pre></td>
<td><pre><span>sampler</span><span>.</span><span>save_anim_gif</span><span>(</span><span>z1</span><span>,</span> <span>z2</span><span>,</span> <span>'output.gif'</span><span>)</span><span>
</span></pre></td>
</tr></tbody></table></code></pre></figure>

</figure>

<h4 id="output-3">Output</h4>

<center>

</center>

<h3 id="colour-images">Colour Images</h3>

<p>If we reset the IPython session and run with this setting to set <code>c_dim</code> to 3, we can also generate colour images as well</p>

<figure>

<figure><pre><code data-lang="python"><table><tbody><tr>
<td><pre>1
2</pre></td>
<td><pre><span>%</span><span>run</span> <span>-</span><span>i</span> <span>sampler</span><span>.</span><span>py</span>
<span>sampler</span> <span>=</span> <span>Sampler</span><span>(</span><span>z_dim</span> <span>=</span> <span>8</span><span>,</span> <span>c_dim</span> <span>=</span> <span>3</span><span>,</span> <span>scale</span> <span>=</span> <span>10.0</span><span>,</span> <span>net_size</span> <span>=</span> <span>32</span><span>)</span><span>
</span></pre></td>
</tr></tbody></table></code></pre></figure>

</figure>

<p>We can get the CPPN to output 3 values for each pixel location, defining RGB value for each pixel.  However, I personally found the black and white version more elegant for generative artworks using random weights.</p>

<center>

</center>

<h3 id="modifying-network-architecture">Modifying Network Architecture</h3>

<p>We can also modify the <code>generate</code> methid inside <code>model.py</code> to change the architecture of the CPPN network.</p>

<p>The default architecture is a few layers of hyperbolic tangent functions</p>

<figure>

<figure><pre><code data-lang="python"><table><tbody><tr>
<td><pre>1
2
3
4</pre></td>
<td><pre><span>H</span> <span>=</span> <span>tf</span><span>.</span><span>nn</span><span>.</span><span>tanh</span><span>(</span><span>U</span><span>)</span>
<span>for</span> <span>i</span> <span>in</span> <span>range</span><span>(</span><span>3</span><span>):</span>
  <span>H</span> <span>=</span> <span>tf</span><span>.</span><span>nn</span><span>.</span><span>tanh</span><span>(</span><span>fully_connected</span><span>(</span><span>H</span><span>,</span> <span>net_size</span><span>,</span> <span>'g_tanh_'</span><span>+</span><span>str</span><span>(</span><span>i</span><span>)))</span>
<span>output</span> <span>=</span> <span>tf</span><span>.</span><span>sigmoid</span><span>(</span><span>fully_connected</span><span>(</span><span>H</span><span>,</span> <span>self</span><span>.</span><span>c_dim</span><span>,</span> <span>'g_final'</span><span>))</span><span>
</span></pre></td>
</tr></tbody></table></code></pre></figure>

</figure>

<h4 id="examples">Examples</h4>
<center>

</center>

<p>Alternatively, we can mix in softplus layers with the tanh layers</p>

<figure>

<figure><pre><code data-lang="python"><table><tbody><tr>
<td><pre>1
2
3
4
5</pre></td>
<td><pre><span>H</span> <span>=</span> <span>tf</span><span>.</span><span>nn</span><span>.</span><span>tanh</span><span>(</span><span>U</span><span>)</span>
<span>H</span> <span>=</span> <span>tf</span><span>.</span><span>nn</span><span>.</span><span>softplus</span><span>(</span><span>fully_connected</span><span>(</span><span>H</span><span>,</span> <span>net_size</span><span>,</span> <span>'g_softplus_1'</span><span>))</span>
<span>H</span> <span>=</span> <span>tf</span><span>.</span><span>nn</span><span>.</span><span>tanh</span><span>(</span><span>fully_connected</span><span>(</span><span>H</span><span>,</span> <span>net_size</span><span>,</span> <span>'g_tanh_2'</span><span>))</span>
<span>H</span> <span>=</span> <span>tf</span><span>.</span><span>nn</span><span>.</span><span>softplus</span><span>(</span><span>fully_connected</span><span>(</span><span>H</span><span>,</span> <span>net_size</span><span>,</span> <span>'g_softplus_2'</span><span>))</span>
<span>output</span> <span>=</span> <span>tf</span><span>.</span><span>sigmoid</span><span>(</span><span>fully_connected</span><span>(</span><span>H</span><span>,</span> <span>self</span><span>.</span><span>c_dim</span><span>,</span> <span>'g_final'</span><span>))</span><span>
</span></pre></td>
</tr></tbody></table></code></pre></figure>

</figure>

<h4 id="examples-1">Examples</h4>
<center>

</center>

<p>We can even mix in sinusoidal activations</p>

<figure>

<figure><pre><code data-lang="python"><table><tbody><tr>
<td><pre>1
2
3
4
5</pre></td>
<td><pre><span>H</span> <span>=</span> <span>tf</span><span>.</span><span>nn</span><span>.</span><span>tanh</span><span>(</span><span>U</span><span>)</span>
<span>H</span> <span>=</span> <span>tf</span><span>.</span><span>nn</span><span>.</span><span>softplus</span><span>(</span><span>fully_connected</span><span>(</span><span>H</span><span>,</span> <span>net_size</span><span>,</span> <span>'g_softplus_1'</span><span>))</span>
<span>H</span> <span>=</span> <span>tf</span><span>.</span><span>nn</span><span>.</span><span>tanh</span><span>(</span><span>fully_connected</span><span>(</span><span>H</span><span>,</span> <span>net_size</span><span>,</span> <span>'g_tanh_2'</span><span>))</span>
<span>H</span> <span>=</span> <span>tf</span><span>.</span><span>nn</span><span>.</span><span>softplus</span><span>(</span><span>fully_connected</span><span>(</span><span>H</span><span>,</span> <span>net_size</span><span>,</span> <span>'g_softplus_2'</span><span>))</span>
<span>output</span> <span>=</span> <span>0.5</span> <span>*</span> <span>tf</span><span>.</span><span>sin</span><span>(</span><span>fully_connected</span><span>(</span><span>H</span><span>,</span> <span>self</span><span>.</span><span>c_dim</span><span>,</span> <span>'g_final'</span><span>))</span> <span>+</span> <span>0.5</span><span>
</span></pre></td>
</tr></tbody></table></code></pre></figure>

</figure>

<h4 id="examples-2">Examples</h4>
<center>

</center>

<p>I have also experimented with absolute functions, square roots, gaussian activations (self implemented), and residual / ladder type structures.</p>

<p>The <code>CPPN</code> class is fairly generic, and has the capability to generate a large batch of images at the same time.  Batch processing is really useful for GPU-based training of a neural network in the future when we want to get CPPN to generate something interesting for us.</p>

<h2 id="conclusions">Conclusions</h2>

<p>The TensorFlow library provides an efficient way to perform vector computations, and can be used in tasks that don’t require gradient computations.  But it means whatever work can be extended in the future to take advantage of the deep learning functionality.  Just like this example, we can then use this code and try to train our CPPN to do interesting things like draw certain class of images, fonts, or numerical digits of different styles, in high resolution.  I will write about this in the upcoming blog post, so please stay tuned.</p>

</section>

    
  
  </div><br>
<br>
<br>
<br>
AI|ML|DM|IR|PR<br>
<br>
via blog.otoro.net http://blog.otoro.net<br>
<br>
October 31, 2018 at 03:39PM
